In [17]:
import pandas as pd

# Работа с Pandas

1. Загрузите ваш DataFrame в переменную df и выведете на экран первые пять строк 

In [18]:
df = pd.read_csv('./data/googleplaystore.csv')
df.iloc[0:5]

App        Category  Rating  \
0     Photo Editor & Candy Camera & Grid & ScrapBook  ART_AND_DESIGN     4.1   
1                                Coloring book moana  ART_AND_DESIGN     3.9   
2  U Launcher Lite – FREE Live Cool Themes, Hide ...  ART_AND_DESIGN     4.7   
3                              Sketch - Draw & Paint  ART_AND_DESIGN     4.5   
4              Pixel Draw - Number Art Coloring Book  ART_AND_DESIGN     4.3   

  Reviews  Size     Installs  Type  Price Content Rating  \
0     159   19M      10,000+  Free    0.0       Everyone   
1     967   14M     500,000+  Free    0.0       Everyone   
2   87510  8.7M   5,000,000+  Free    0.0       Everyone   
3  215644   25M  50,000,000+  Free    0.0           Teen   
4     967  2.8M     100,000+  Free    0.0       Everyone   

                      Genres      Last Updated         Current Ver  \
0               Art & Design   January 7, 2018               1.0.0   
1  Art & Design;Pretend Play  January 15, 2018               2.0.0   
2               Art & Design    August 1, 2018               1.2.4   
3               Art & Design      June 8, 2018  Varies with device   
4    Art & Design;Creativity     June 20, 2018                 1.1   

    Android Ver  
0  4.0.3 and up  
1  4.0.3 and up  
2  4.0.3 and up  
3    4.2 and up  
4    4.4 and up

2. Отобразите количество строк и столбцов в DataFrame

In [19]:
df.shape

(10841, 13)

3. Выведете основную информацию о DataFrame методом info(). Проанализируйте есть ли пропуски в столбцах. Напишите вывод о том, какие столбцы содержат пропуски

In [20]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10841 entries, 0 to 10840
Data columns (total 13 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   App             10841 non-null  object 
 1   Category        10841 non-null  object 
 2   Rating          9367 non-null   float64
 3   Reviews         10841 non-null  object 
 4   Size            10841 non-null  object 
 5   Installs        10841 non-null  object 
 6   Type            10840 non-null  object 
 7   Price           10841 non-null  float64
 8   Content Rating  10840 non-null  object 
 9   Genres          10841 non-null  object 
 10  Last Updated    10841 non-null  object 
 11  Current Ver     10833 non-null  object 
 12  Android Ver     10838 non-null  object 
dtypes: float64(2), object(11)
memory usage: 1.1+ MB


<h1>Вывод</h1>
<p>Видем, что индекс диапазона: 10841 записей - это строка без пропусков.</p>
<p>Сравниваем визуально, например:</p>
<p>App 10841 = 10841 нет пропусков</p>
<p>Rating 9367 < 10841 пропуски есть</p>
<p>Вывод: столбцы Rating, Type, Content, Rating, Current Ver, Android Ver содержат пропуски</p>

4. Удалите или замените пропуски известными вам методами. (Подсказка: `dropna()` или `fillna()`, выбор того или иного метода должен быть аргументирован).

In [21]:
df.Rating = df.Rating.fillna(df.Rating.median())
df.dropna(inplace=True)
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 10829 entries, 0 to 10840
Data columns (total 13 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   App             10829 non-null  object 
 1   Category        10829 non-null  object 
 2   Rating          10829 non-null  float64
 3   Reviews         10829 non-null  object 
 4   Size            10829 non-null  object 
 5   Installs        10829 non-null  object 
 6   Type            10829 non-null  object 
 7   Price           10829 non-null  float64
 8   Content Rating  10829 non-null  object 
 9   Genres          10829 non-null  object 
 10  Last Updated    10829 non-null  object 
 11  Current Ver     10829 non-null  object 
 12  Android Ver     10829 non-null  object 
dtypes: float64(2), object(11)
memory usage: 1.2+ MB


<h2>Аргументация</h2>
Пропущенных данных из Rating слишком много, поэтому заполняем медианой, которая берёт центральное значение и не зависит от крайних значений, а остальные пропуски удаляем, т.к. их мало и они не повлияют на статистику

5. Проверьте нет ли пропусков после произведенной обработки

In [22]:
df.isna().sum()

App               0
Category          0
Rating            0
Reviews           0
Size              0
Installs          0
Type              0
Price             0
Content Rating    0
Genres            0
Last Updated      0
Current Ver       0
Android Ver       0
dtype: int64

6. Выведите на экран минимум и максимум из столбца `Price`:

In [23]:
print("мин:", df.Price.min())
print("макс:", df.Price.max())

мин: 0.0
макс: 400.0


7. Выведите на экран медиану и среднее арифметическое столбцов:
 - Rating
 - Reviews

In [24]:
df.Reviews = df.Reviews.str.replace("M", "000000").astype(int)
#Одно из значений содержит букву "M" вместо цифр, поэтому заменяем ее на 000000 и устанавливаем числовой тип

In [25]:
print("Rating - Медиана:", df.Rating.median())
print("Rating - Среднее:", df.Rating.mean())

print("Reviews - Медиана:", df.Reviews.median())
print("Reviews - Среднее:", df.Reviews.mean())

Rating - Медиана: 4.3
Rating - Среднее: 4.206510296426262
Reviews - Медиана: 2100.0
Reviews - Среднее: 444601.76673746423


8. Выведите на экран все уникальные значения категориального столбца Genres

In [26]:
df.Genres.unique()

array(['Art & Design', 'Art & Design;Pretend Play',
       'Art & Design;Creativity', 'Art & Design;Action & Adventure',
       'Auto & Vehicles', 'Beauty', 'Books & Reference', 'Business',
       'Comics', 'Comics;Creativity', 'Communication', 'Dating',
       'Education;Education', 'Education', 'Education;Creativity',
       'Education;Music & Video', 'Education;Action & Adventure',
       'Education;Pretend Play', 'Education;Brain Games', 'Entertainment',
       'Entertainment;Music & Video', 'Entertainment;Brain Games',
       'Entertainment;Creativity', 'Events', 'Finance', 'Food & Drink',
       'Health & Fitness', 'House & Home', 'Libraries & Demo',
       'Lifestyle', 'Lifestyle;Pretend Play',
       'Adventure;Action & Adventure', 'Arcade', 'Casual', 'Card',
       'Casual;Pretend Play', 'Action', 'Strategy', 'Puzzle', 'Sports',
       'Music', 'Word', 'Racing', 'Casual;Creativity',
       'Casual;Action & Adventure', 'Simulation', 'Adventure', 'Board',
       'Trivia', 'Role 

9. Сгруппируйте данные по столбцу `Genres` и посчитайте для каждого жанра средний и медианный рейтинг. Представьте результат в виде нового DataFrame,который будет сохранен в переменную `grouped_df`, где одна колонка будет содержать названия жарнов, а две другие буду содержать в себе средний и медианный рейтинги. Напишите краткий вывод.

In [27]:
grouped_df = df.groupby(by = "Genres", as_index=False).agg(Средний_Rating = ('Rating','mean'), Медианный_Rating = ('Rating','median'))
grouped_df

Genres  Средний_Rating  Медианный_Rating
0                                   Action        4.285753               4.3
1                Action;Action & Adventure        4.311765               4.3
2                                Adventure        4.184000               4.3
3             Adventure;Action & Adventure        4.423077               4.5
4                    Adventure;Brain Games        4.600000               4.6
..                                     ...             ...               ...
114                Video Players & Editors        4.084393               4.3
115     Video Players & Editors;Creativity        4.100000               4.1
116  Video Players & Editors;Music & Video        4.000000               4.0
117                                Weather        4.248780               4.3
118                                   Word        4.406897               4.4

[119 rows x 3 columns]

<h1>Вывод</h1>
<p>Из датасета мы вывели таблицу жанров приложений, в которой наглядно видно небольшое различие между средним и медианным рейтингом, а так же, что почти все жанры имеют рейтинг выше 4</p>

10. Из получившегося датафрейма выведите на экран только те жанры, чей медианный рейтинг больше 4.5. Напишите краткий вывод.

In [28]:
grouped_df[grouped_df["Медианный_Rating"] > 4.5]

Genres  Средний_Rating  Медианный_Rating
4          Adventure;Brain Games        4.600000               4.6
11       Art & Design;Creativity        4.400000               4.7
18            Board;Pretend Play        4.800000               4.8
35             Comics;Creativity        4.800000               4.8
55      Entertainment;Creativity        4.533333               4.6
64    Health & Fitness;Education        4.700000               4.7
74           Music;Music & Video        4.533333               4.6
87              Puzzle;Education        4.600000               4.6
105  Strategy;Action & Adventure        4.600000               4.6

<h1>Вывод</h1>
<p>Из таблицы мы можем увидеть, наиболее популярные жанры среди пользователей</p>

11. Выведете кол-во дубликатов в основном DataFrame (df) и удалите их 

In [29]:
print("кол-во дубликатов:", df.duplicated().sum())
df.drop_duplicates(inplace=True)


кол-во дубликатов: 483


12. Посчитатайте какое количество приложений содержится в каждом жанре (value_counts).В получившемся Series выведите на экран 10 строк с наибольшим количеством приложений отсортированных по убыванию.

In [31]:
top_genres = df.Genres.value_counts()
print(top_genres.head(10))

Genres
Tools              840
Entertainment      587
Education          526
Business           427
Medical            408
Productivity       407
Personalization    386
Lifestyle          372
Communication      366
Sports             364
Name: count, dtype: int64
